feat(docs): Independent ticketing system tutorial (#4754)
iotaledger/iota: docs/content/developer/getting-started/independent-ticketing-system/package.mdx

Reviewer: Can you mention unit tests you added in `package.mdx` please?

Author: absolutely, sure.

---
title: Package
description: Tutorial for Independent Ticketing System Package
tags:
  - explanation
  - tutorial
  - move
  - nft
  - sdk
  - cli
---

# Independent Ticketing System Tutorial

> In this tutorial, we will build an Independent Ticketing System utilizing the Move language for blockchain contracts and the [IOTA dApp kit](../../../ts-sdk/dapp-kit) for the frontend application. The project begins with creating a Move package, building and deploying it on the testnet network. Once the backend contract is operational, we will develop the frontend dApp. 

Reviewer 2: Please add links to the Create a Move Package header, and the FE doc

Author: Sure.

> ## Sequence Diagram 

Reviewer: It would be clearer to differentiate between ticket/event owner and ticket buyer in the sequence diagram I think

Author: sure

> 

Reviewer: Can you add a small paragraph explaining the overall flow?

Author: sure, thanks for the suggestion

    NFTContract ->> NFTContract: Validate sender is buyer
    NFTContract ->> Blockchain: Split coin for royalty and seller payment
    NFTContract ->> Blockchain: Transfer royalty fee to creator
    NFTContract ->> Blockchain: Transfer payment to seller
    NFTContract ->> Blockchain: Public transfer of TicketNFT to buyer
    NFTContract ->> NFTContract: Delete InitiateResale object

    User ->> NFTContract: burn(nft, ctx)
    NFTContract ->> NFTContract: Validate sender is NFT owner
    NFTContract ->> Blockchain: Delete NFT object

    User ->> NFTContract: whitelist_buyer(user, nft)
    NFTContract ->> NFTContract: Add user to NFT whitelist

    User ->> NFTContract: create_TicketNFT(name, creator, owner, ... , ctx)
    NFTContract ->> NFTContract: Create a new TicketNFT (Test purpose)

    User ->> NFTContract: create_TotalSeat(value, ctx)
    NFTContract ->> NFTContract: Create TotalSeat object (Test purpose)
```
## Prerequisites

- [Node.js](https://nodejs.org/en) >= v20.18.0
- [npx](https://www.npmjs.com/package/npx) >= 10.8.2
- [iota](https://github.com/iotaledger/iota/releases) >= 0.7.3

## [Create a Move Package](../create-a-package.mdx)

> Run the following command to create a Move package:

Reviewer 2: Please add a link to /developer/getting-started/create-a-package

Author: Sure.

Reviewer 2: ```suggestion
Run the following command to [create a Move package](../create-a-package.mdx):
```
We should use relative links whenever possible.

Reviewer 2: Why is this resolved?

Author: My bad, check now.

> Run the following command to create a Move package:

Reviewer 2: Please add a link to /developer/getting-started/create-a-package

Author: Sure.

```bash
iota move new independent_ticketing_system
```

## Package Overview

### [`init`](https://github.com/iota-community/independent-ticketing-system/blob/6c6cd27bbf402dc3e1a219a2d6e5556eb968bb59/independent_ticketing_system/sources/independent_ticketing_system.move#L71-L85)

> In MOVE, the `init` function only runs once at the time of package publication. Read more: [`init`](../../iota-101/move-overview/init)

Reviewer 2: ```suggestion
In Move, the [`init`](../../iota-101/move-overview/init) function only runs once at package publication.
```

> In MOVE, the `init` function only runs once at the time of package publication. Read more: [`init`](../../iota-101/move-overview/init)

Reviewer 2: ```suggestion
In Move, the [`init`](../../iota-101/move-overview/init) function only runs once at the time of package publication.
```

The `init` function will create three [`shared_object`](../../iota-101/objects/object-ownership/shared.mdx) type objects: 
- [`Creator`](https://github.com/iota-community/independent-ticketing-system/blob/6c6cd27bbf402dc3e1a219a2d6e5556eb968bb59/independent_ticketing_system/sources/independent_ticketing_system.move#L20-L23) will store the address of the package creator.
- [`TotalSeat`](https://github.com/iota-community/independent-ticketing-system/blob/6c6cd27bbf402dc3e1a219a2d6e5556eb968bb59/independent_ticketing_system/sources/independent_ticketing_system.move#L25-L28) - it will contain the total number of seat for the event which will update as per minting the tickets.

> - [`AvailableTicketsToBuy`](https://github.com/iota-community/independent-ticketing-system/blob/6c6cd27bbf402dc3e1a219a2d6e5556eb968bb59/independent_ticketing_system/sources/independent_ticketing_system.move#L38-L41) - it will contains the array of all the nft which are minted by the  creator and are available to buy. To buy tickets first user needs to get whitelisted. 

Reviewer 2: ```suggestion
- [`AvailableTicketsToBuy`](https://github.com/iota-community/independent-ticketing-system/blob/6c6cd27bbf402dc3e1a219a2d6e5556eb968bb59/independent_ticketing_system/sources/independent_ticketing_system.move#L38-L41) - it will contain the array of all the NFT which are minted by the creator and are available to buy. To buy tickets, the first user needs to be whitelisted. 
```

- This function burns the NFT, removing the NFT object from the owner's account.  
- Only the NFT's owner is authorized to burn the NFT.

```move reference
https://github.com/iota-community/independent-ticketing-system/blob/6c6cd27bbf402dc3e1a219a2d6e5556eb968bb59/independent_ticketing_system/sources/independent_ticketing_system.move#L235-L250
```

### [`whitelist_buyer`](https://github.com/iota-community/independent-ticketing-system/blob/6c6cd27bbf402dc3e1a219a2d6e5556eb968bb59/independent_ticketing_system/sources/independent_ticketing_system.move#L258-L260)

- This function allows the owner of an NFT to whitelist a user.  
- Once whitelisted, the user becomes eligible to purchase the NFT.

```move reference
https://github.com/iota-community/independent-ticketing-system/blob/6c6cd27bbf402dc3e1a219a2d6e5556eb968bb59/independent_ticketing_system/sources/independent_ticketing_system.move#L258-L260
```

See the full example of the contract on [github](https://github.com/iota-community/independent-ticketing-system/blob/main/independent_ticketing_system/sources/independent_ticketing_system.move) .

### [Publish the Package](https://docs.iota.org/developer/getting-started/publish)

> Publish the package to the IOTA testnet using the following command:

Reviewer 2: Please add a link to /developer/getting-started/publish

Author: Sure.

Reviewer 2: ```suggestion
[Publish](../publish.mdx) the package to the IOTA Testnet using the following command:
```

Reviewer 2: Why is this resolved?

Author: My bad, check now.

> Publish the package to the IOTA testnet using the following command:

Reviewer 2: Please add a link to /developer/getting-started/publish

Author: Sure.

    NFTContract ->> Blockchain: Split coin for royalty and seller payment
    NFTContract ->> Blockchain: Transfer royalty fee to creator
    NFTContract ->> Blockchain: Transfer payment to seller
    NFTContract ->> Blockchain: Public transfer of TicketNFT to buyer
    NFTContract ->> NFTContract: Delete InitiateResale object

    User ->> NFTContract: burn(nft, ctx)
    NFTContract ->> NFTContract: Validate sender is NFT owner
    NFTContract ->> Blockchain: Delete NFT object

    User ->> NFTContract: whitelist_buyer(user, nft)
    NFTContract ->> NFTContract: Add user to NFT whitelist

    User ->> NFTContract: create_TicketNFT(name, creator, owner, ... , ctx)
    NFTContract ->> NFTContract: Create a new TicketNFT (Test purpose)

    User ->> NFTContract: create_TotalSeat(value, ctx)
    NFTContract ->> NFTContract: Create TotalSeat object (Test purpose)
```
## Prerequisites

- [Node.js](https://nodejs.org/en) >= v20.18.0
- [npx](https://www.npmjs.com/package/npx) >= 10.8.2
- [iota](https://github.com/iotaledger/iota/releases) >= 0.7.3

## Create a Move Package

Run the following command to [create a Move package](/developer/getting-started/create-a-package):

```bash

> iota move new independent_ticketing_system

Reviewer 3: Maybe easy for newcomers to miss that they aren't in the correct folder

```suggestion
iota move new independent_ticketing_system && cd independent_ticketing_system
```

```

## Package Overview

### [`init`](https://github.com/iota-community/independent-ticketing-system/blob/6c6cd27bbf402dc3e1a219a2d6e5556eb968bb59/independent_ticketing_system/sources/independent_ticketing_system.move#L71-L85)

In Move, the [`init`](../../iota-101/move-overview/init) function only runs once at package publication.
The `init` function will create three [`shared_object`](../../iota-101/objects/object-ownership/shared.mdx) type objects: 

> - [`Creator`](https://github.com/iota-community/independent-ticketing-system/blob/6c6cd27bbf402dc3e1a219a2d6e5556eb968bb59/independent_ticketing_system/sources/independent_ticketing_system.move#L20-L23) will store the address of the package creator.

Reviewer 3: This seems a little bit like an anti pattern. Shouldn't we use a cap for this? You don't even need to store the creator address. You could add the cap ID to the NFT if really needed. That way people could get additional infos like when the event was created for example 🤔

Author: Great Idea, I implemented the cap, check now.
Thanks for the suggestion.

- [`TotalSeat`](https://github.com/iota-community/independent-ticketing-system/blob/6c6cd27bbf402dc3e1a219a2d6e5556eb968bb59/independent_ticketing_system/sources/independent_ticketing_system.move#L25-L28) - it will contain the total number of seat for the event which will update as per minting the tickets.

> - [`AvailableTicketsToBuy`](https://github.com/iota-community/independent-ticketing-system/blob/6c6cd27bbf402dc3e1a219a2d6e5556eb968bb59/independent_ticketing_system/sources/independent_ticketing_system.move#L38-L41) - it will contain the array of all the NFT which are minted by the creator and are available to buy. To buy tickets, the first user needs to be whitelisted. 

Reviewer 3: Why not unify it into one shared object? An event object for example.

Author: Sure.
Thanks for the suggestion.

> 

Reviewer 2: I think Smart Contract or Move Package make a more descriptive sidebar label. Btw, you can use the [`sidebar_label` ](https://docusaurus.io/docs/api/plugins/@docusaurus/plugin-content-docs#sidebar_label) attribute for this, or set it in the sidebar file as well.

Author: sure, thanks for suggestion.

> # Independent Ticketing System Tutorial

Reviewer 2: Just in case, this header is used over the title attribute in the yml. We may want to say something about this being the SC part of the tutorial.

Author: Sure.

- [npx](https://www.npmjs.com/package/npx) >= 10.8.2
- [iota](https://github.com/iotaledger/iota/releases) >= 0.7.3

> ## [Create a Move Package](/developer/getting-started/create-a-package)

Reviewer 2: ```suggestion
## Create a Move Package
```

See the full example of the contract on [github](https://github.com/iota-community/independent-ticketing-system/blob/main/independent_ticketing_system/sources/independent_ticketing_system.move) .

> ### Write unit tests

Reviewer 2: ```suggestion
### Write Unit Tests
```

After creating the module write the test for all the functions. Move the below code to the `independent_ticketing_system_test.move` file:

```move reference 
https://github.com/iota-community/independent-ticketing-system/blob/794492c7710358da53a9c01a85011f42eaea5500/independent_ticketing_system/tests/independent_ticketing_system_test.move
```

> ### [Publish the Package](/developer/getting-started/publish)

Reviewer 2: ```suggestion
### Publish the Package
```
If possible, we should avoid adding links within headers.

Author: Will definitely do it.
